ql/experimental/callablebonds/callablebond.cpp

```diff
@@ -108,6 +109,270 @@ namespace QuantLib {
     }
 
 
+    /* Expose a mechanism to add spread the engine parameters and
+       clean clean up when the object is destroyed
+     */
+    class EngSpreadHelper {
+        CallableBond::arguments*  args;
+        double ospread;
+    public:
+        explicit EngSpreadHelper(CallableBond::arguments* args):
+            args(args),
+            ospread(args->spread)
+        {
+        }
+
+        void setSpread(double s)
+        {
+            args->spread=s;
+        }
+
+        ~EngSpreadHelper()
+        {
+            setSpread(ospread);
+        }
+    };
+
+    /* Functional interface to take a spread and return NPV of
+       spreaded callable bond */
+    class NPVSpreadHelper :
+        public std::unary_function<Real, Real>
+    {
+        CallableBond& bond_;
+        EngSpreadHelper &e;
+    public:
+        NPVSpreadHelper(CallableBond& bond,
+                        EngSpreadHelper& e):
+            bond_(bond),
+            e(e)
+        {
+        }
+        Real operator()(Real x) const
+        {
+            e.setSpread(x);
+            bond_.recalculate();
+            return bond_.NPV();
```

Comment:
This relies on the fact that `bond_.setupArguments()`, which is called inside `recalculate`, won't overwrite the `spread` data member in the arguments.  It's kind of fragile.  I would use the engine instead of the bond and write something like

    arguments_.spread = x;
    engine_.calculate();
    return results_.value;

The above requires that the rest of the arguments are initialized; you can do that by calling `bond.setupArguments(arguments_)`, but that can be done at initialization instead of at each call.

Reply:
Or you can use

    engine_.calculateWithSpread(x);
    return results_.value;

and avoid adding `spread` to the arguments, where it probably doesn't belong.